api/app_sph_lms/api/serializer/course_serializer.py

```diff
@@ -74,6 +77,46 @@ def create(self, validated_data):
         course.category.set(categories_data)
         return course
 
+    def update(self, instance, validated_data):
+        user = self.context['request'].user
+
+        if not user.is_authenticated or \
+                user.role.title not in ['Trainer', 'Admin']:
+            raise PermissionDenied(
+                "Only authenticated Trainers and Admins can update a course."
+            )
+
+        categories_data = validated_data.pop('category')
+        lessons_data = validated_data.pop('lessons')
+
+        for attr, value in validated_data.items():
+            setattr(instance, attr, value)
+        instance.save()
+
+        existing_lesson_ids = [lesson.id for lesson in instance.lessons.all()]
+
+        for lesson_data in lessons_data:
+            lesson_id = lesson_data.get('id')
+
+            if lesson_id in existing_lesson_ids:
+                lesson = Lesson.objects.get(id=lesson_id, course=instance)
+                for attr, value in lesson_data.items():
+                    setattr(lesson, attr, value)
+                lesson.save()
+                existing_lesson_ids.remove(lesson_id)
+            else:
+                lesson_data['course'] = instance
+                Lesson.objects.create(**lesson_data)
+
+        Lesson.objects.filter(
+            id__in=existing_lesson_ids,
+            course=instance
+        ).delete()
+
+        instance.category.set(categories_data)
```

Comment:
Seems like you are still using the old course model, and you are able to save the categories in course instance by using this. Please double check the new ERD and update the course model accordingly, including the saving of course categories, 
store it in CourseCategory table